### ISSUE TITLE
Cannot invoke resources with POST, PUT and DELETE in the API after changing the API endpoint URL via etcd

### ISSUE THREAD

**Issue by NatashaWso2** [wso2/product-microgateway#923]
### Description:
I have enabled service discovery using etcd on WSO2 API Microgateway as provided in [1]. I have used the attached Open API definition to create the Microgateway project : 
[bookstore-swagger.txt](https://github.com/wso2/product-microgateway/files/4029842/bookstore-swagger.txt)

Initially the production endpoint URL will be pointed to **http://localhost:9191/bookattic**. After starting the WSO2 Microgateway I can invoke all the resources defined in the API definition using a REST client without any issue. When I change the API endpoint URL via etcd [2] to the new URL  i.e. to **http://localhost:9195/bookhaven/**, the new endpoint change becomes visible to the Microgateway after sometime. However, I can only successfully invoke the GET resources specified in the API definition. When I try to invoke the other resources, I get the following error:

_**{
           "fault": {
               "code": "101503",
              "message": "Runtime Error",
             "description": "Error connecting to the back end"
          }
}**_

However when I restart the Microgateway, then I can successfully invoke all the resources defined in the API.

[1] [https://docs.wso2.com/display/MG300/Service+Discovery#ServiceDiscovery-Enablingservicediscovery](https://docs.wso2.com/display/MG300/Service+Discovery#ServiceDiscovery-Enablingservicediscovery)
[2] [https://docs.wso2.com/display/MG300/Service+Discovery#ServiceDiscovery-ChangingtheAPIendpointURLviaetcd](https://docs.wso2.com/display/MG300/Service+Discovery#ServiceDiscovery-ChangingtheAPIendpointURLviaetcd)

### Affected Product Version:
WSO2 Microgateway 3.0.2

### Environment details (with versions):
- OS: Ubuntu 19.04

**Comment by praminda** [comment]
@NatashaWso2 What are the backends you are using?

**Comment by praminda** [comment]
@NatashaWso2 Can you please attach the project to the issue

**Comment by NatashaWso2** [comment]
@praminda I have compressed and attached the project and the backends

[mgw-project.zip](https://github.com/wso2/product-microgateway/files/4035019/mgw-project.zip)

**Comment by praminda** [comment]
Hi @NatashaWso2
Looks like issue is reproducible. I'm looking into the cause of the issue ATM

**Comment by praminda** [comment]
Ok, found the issue.
In following code line `@index` is used to define the endpoint index. However for http endpoints (when the endpoint is not load balancer or failover etc.), `@index` doesn't mean anything as this code line is not inside a loop. In this case Mustache was resolving the value of `@index` to 0 and sometimes 1. Which cause the etcd implementation to fail when trying to find the configured endpoint for some resources.
https://github.com/wso2/product-microgateway/blob/efde86607bbfaa7d5dd86507a4c9651625f56b3d/components/micro-gateway-cli/src/main/resources/templates/httpEndpointReInit.mustache#L4

As a temporary resolution until we fix the issue in a future release you can replace @index part from above code line with 0.
Result should look like,
```ballerina
(<string>gateway:etcdUrls[<string> gateway:retrieveConfig("{{name}}_{{endpointUrlType}}_0_etcdKey","{{etcdKey}}")], config = { {{>http2}}{{>caching}}{{>basicAuth}}{{>secureSocket}} });
```
This file can be found in `<MGW_TOOLKIT>/resources/templates/httpEndpointReInit.mustache`

**Comment by Rajith90** [comment]
Lets fix this in mater branch and 3.0.x branch as well as well
